fusion-core/src/types.js

```diff
@@ -10,7 +10,7 @@ import type {Context as KoaContext} from 'koa';
 
 export type Token<T> = {
   (): T,
-  optional: () => void | T,
+  optional: Token<void | T>,
```

Comment:
What are the implications here if we instead did `optional: ?Token<T>` which matches the `create-token.js` implementation more closely:

https://github.com/fusionjs/fusionjs/blob/73914f34f02187c54c9da51e5d7cc48e94aedf74/fusion-core/src/create-token.js#L16-L35

Reply:
I actually think the `create-token.js` implementation is wrong, since it's optional but not nullable. If I try to make the type nullable instead, there will be a ton of other errors (example where a prior check would now fail: https://github.com/fusionjs/fusionjs/blob/master/fusion-plugin-styletron-react/src/browser.js#L42).

Reply:
Good point.  Given that `?Token<T>` is equivalent to `null | void | Token<T>`, how about `optional: void | Token<T>` instead?

Reply:
Discussed offline:

Larger issue is that the types in `types.js` and the implementation are inconsistent.  Changes that make `optional` potentially `null` or `void` cause regressions that we would need to fix.

Current proposal keeps us relatively close to the runtime behavior, but allows some odd behaviors (e.g. `token.optional.optional.optional...`).

Reply:
Discussed offline - the north star solution here is a bit trickier to get to, whereas the current approach at least mirrors the old logic to avoid breaking other things.

Idea: make the `optional` field non-optional. Have the required token's `optional` point to the optional token, and have the optional token's `optional` point to itself. Someone more familiar with the codebase should probably play around with that, though.